systems/unix/system.go

```diff
@@ -989,55 +988,61 @@ func (s *System) SockAddressInfo(ctx context.Context, name, service string, hint
 			ip = net.IPv4zero
 		}
 	}
-	if ip != nil {
-		results = results[:1]
-		results[0] = wasi.AddressInfo{}
+
+	makeAddressInfo := func(ip net.IP, port int) wasi.AddressInfo {
+		addrInfo := wasi.AddressInfo{
+			Flags:      hints.Flags,
+			SocketType: hints.SocketType,
+			Protocol:   hints.Protocol,
+		}
 		if ipv4 := ip.To4(); ipv4 != nil {
 			inet4Addr := &wasi.Inet4Address{Port: port}
 			copy(inet4Addr.Addr[:], ipv4)
-			results[0].Address = inet4Addr
+			addrInfo.Family = wasi.InetFamily
+			addrInfo.Address = inet4Addr
 		} else {
 			inet6Addr := &wasi.Inet6Address{Port: port}
 			copy(inet6Addr.Addr[:], ip)
-			results[0].Address = inet6Addr
+			addrInfo.Family = wasi.Inet6Family
+			addrInfo.Address = inet6Addr
 		}
+		return addrInfo
+	}
+
+	if ip != nil {
+		results[0] = makeAddressInfo(ip, port)
 		return 1, wasi.ESUCCESS
 	}
 
-	ips, err := net.LookupIP(name)
+	// LookupIP requires the network to be one of "ip", "ip4", or "ip6".
+	switch network {
+	case "tcp", "udp":
+		network = "ip"
+	case "tcp4", "udp4":
+		network = "ip4"
+	case "tcp6", "udp6":
+		network = "ip6"
+	}
+
+	ips, err := net.DefaultResolver.LookupIP(ctx, network, name)
 	if err != nil {
 		return 0, wasi.ECANCELED // TODO: better errors on name resolution failure
 	}
 
-	addrs := make([]wasi.AddressInfo, 0, 16)
+	addrs4 := make([]wasi.AddressInfo, 0, 8)
+	addrs6 := make([]wasi.AddressInfo, 0, 8)
```

Comment:
I'm confused about this. Why have two slices? or why not append directly into `results`?

Reply:
There are two slices so that we can put all the IPv4 addresses first; this is a cheaper and simpler way to do the partial ordering that was implemented with a stable sort in the previous version.

`results` is an output parameter, we can't append to the slice, which is why we use `copy` after collecting the addresses.

Reply:
Ah I forgot about the need to have ipv4 first 👍🏻